### PR TITLE
[jvm-packages] Exposed train-time evaluation metrics

### PR DESCRIPTION
**Pull request by superbobry**
They are accessible via `XGBoostModel.summary`. The summary is not serialized with the model and is only available after the training.

### CODE REVIEW

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala [hunk 1]
```diff
@@ -326,24 +330,24 @@ object XGBoost extends Serializable {
...
       val trackerReturnVal = parallelismTracker.execute(tracker.waitFor(0L))
       logger.info(s"Rabit returns with exit code $trackerReturnVal")
-      val model = postTrackerReturnProcessing(trackerReturnVal, boosters, overriddenParams,
-        sparkJobThread, isClsTask)
+      val model = postTrackerReturnProcessing(trackerReturnVal, boostersAndMetrics,
+        overriddenParams, sparkJobThread, isClsTask)
       if (isClsTask){
         model.asInstanceOf[XGBoostClassificationModel].numOfClasses =
-          params.getOrElse("num_class", "2").toString.toInt
+            params.getOrElse("num_class", "2").toString.toInt
```
Comment (sethah) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: Why change this?

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala [hunk 2]
```diff
@@ -352,13 +356,14 @@ object XGBoost extends Serializable {
   }
 
   private def postTrackerReturnProcessing(
-      trackerReturnVal: Int, distributedBoosters: RDD[Booster],
+      trackerReturnVal: Int, distributedBoosters: RDD[(Booster, Map[String, Array[Float]])],
```
Comment (sethah) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: `distributedBoostersAndMetrics`?

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostModel.scala [hunk 3]
```diff
@@ -42,6 +42,21 @@ abstract class XGBoostModel(protected var _booster: Booster)
   extends PredictionModel[MLVector, XGBoostModel] with BoosterParams with Serializable
     with Params with MLWritable {
 
+  private var _trainingSummary: Option[XGBoostTrainingSummary] = None
```
Comment (sethah) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostModel.scala]: why underscore?

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostModel.scala [hunk 4]
```diff
@@ -42,6 +42,21 @@ abstract class XGBoostModel(protected var _booster: Booster)
...
+
+  /**
+   * Returns summary (e.g. train/test objective history) of model on the
+   * training set. An exception is thrown if no summary is available.
+   */
+  def summary: XGBoostTrainingSummary = _trainingSummary.getOrElse {
+    throw new IllegalStateException("No training summary available for this XGBoostModel")
+  }
+
+  def setSummary(summary: XGBoostTrainingSummary): this.type = {
```
Comment (sethah) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostModel.scala]: this is not exposed publicly in Spark. Is there some reason to do it here?

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala [hunk 5]
```diff
@@ -135,10 +135,14 @@ object XGBoost extends Serializable {
       try {
         val numEarlyStoppingRounds = params.get("numEarlyStoppingRounds")
             .map(_.toString.toInt).getOrElse(0)
+        val metrics = Array.tabulate(watches.size)(_ => Array.ofDim[Float](round))
         val booster = SXGBoost.train(watches.train, params, round,
-          watches = watches.toMap, obj = obj, eval = eval,
+          watches.toMap, metrics, obj, eval,
           earlyStoppingRound = numEarlyStoppingRounds)
-        Iterator(booster)
+        val metricsMap = watches.toMap.keys.zipWithIndex
```
Comment (sethah) [inline jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: `val metricsMap = watches.toMap.keys.zip(metrics).toMap`?

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostTrainingSummary.scala [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+
+package ml.dmlc.xgboost4j.scala.spark
+
+class XGBoostTrainingSummary private(
+    val trainObjectiveHistory: Array[Float],
+    val testObjectiveHistory: Option[Array[Float]]
+) extends Serializable
```
[the remark this answers is not in the document]
Comment (superbobry) [reply jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostTrainingSummary.scala]: Done.

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostTrainingSummary.scala [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+object XGBoostTrainingSummary {
```
[the remark this answers is not in the document]
Comment (superbobry) [reply jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostTrainingSummary.scala]: Done.

File: jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostGeneralSuite.scala [hunk 8]
```diff
@@ -95,6 +95,7 @@ class XGBoostGeneralSuite extends FunSuite with PerTest {
       "objective" -> "binary:logistic").toMap
     val xgBoostModel = XGBoost.trainWithRDD(trainingRDD, paramMap, round = 5,
       nWorkers = numWorkers, useExternalMemory = true)
+    xgBoostModel.summary
```
[the remark this answers is not in the document]
Comment (superbobry) [reply jvm-packages/xgboost4j-spark/src/test/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostGeneralSuite.scala]: Done.

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala (hunk 1 above)
Comment (superbobry) [reply jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: Accidental change, will revert.

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala (hunk 5 above)
Comment (superbobry) [reply jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: Good catch! Fixed.

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostModel.scala (hunk 3 above)
Comment (superbobry) [reply jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostModel.scala]: Removed.

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostModel.scala (hunk 4 above)
Comment (superbobry) [reply jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostModel.scala]: Restricted to `spark`.

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostModel.scala (hunk 3 above)
Comment (superbobry) [reply jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostModel.scala]: Actually, it was mimicking the naming in `XGBoostModel`, but I have no strong preferences on this.

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala (hunk 2 above)
Comment (superbobry) [reply jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala]: Done.

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassificationModel.scala [hunk 9]
```diff
@@ -171,9 +171,8 @@ class XGBoostClassificationModel private[spark](
   def numClasses: Int = numOfClasses
 
   override def copy(extra: ParamMap): XGBoostClassificationModel = {
-    val clsModel = defaultCopy(extra).asInstanceOf[XGBoostClassificationModel]
-    clsModel._booster = booster
-    clsModel
+    val newModel = copyValues(new XGBoostClassificationModel(booster), extra)
```
[the remark this answers is not in the document]
Comment (superbobry) [reply jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassificationModel.scala]: Done for both classes.

**Comment by superbobry** [comment]
Note: CI failure on Appveyor is unrelated to the patch.

File: jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassificationModel.scala (hunk 9 above)
[the remark this answers is not in the document]
Comment (CodingCat) [reply jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoostClassificationModel.scala]: ah, sorry for the confusion, it should work

**Comment by superbobry** [comment]
Hey @CodingCat, could you have another look at this please?

**Comment by CodingCat** [comment]
LGTM

**Pull request merged**